README.md

```diff
@@ -489,6 +489,52 @@ This filter provides the capability to enable a feature based on a time window.
 }
 ```
 
+This filter also provides the capability to enable a feature during some recurring time windows. The recurring time windows act as the additional filters of evaluation, supplementing the time window defined by the `Start` and `End`. If any recurring time window filter is specified, the feature will only be enabled when the current time falls between the `Start` and `End`, and also falls within a recurring time window listed in the `Filters`.
+We can see some Crontab expressions are listed under the `Filters`, which specify the recurring time windows. 
+``` JavaScript
+"EnhancedPipeline": {
+    "EnabledFor": [
+        {
+            "Name": "Microsoft.TimeWindow",
+            "Parameters": {
+                "Start": "2023-09-06T00:00:00+08:00",
+                "Filters": [
+                    // 18:00-20:00 on weekdays
+                    "* 18-19 * * Mon-Fri",
+                    // 18:00-22:00 on weekends
+                    "* 18-21 * * Sat,Sun"
+                ]

```

Comment:
Do we really have to support an array of crontab expressions, or is one sufficient for most scenarios? Even if someone really needs more than one, they can always add more Time Window filters with different configurations.

I'm concerned this complicates the usage of the recurring Time Window filter. I can already see this complicates the UI design.

BTW, having a 'Filters' parameter in a feature filter is also confusing.

Reply:
The scenario that turn on a feature flag in 18:00-20:00 on every weekend and 18:00-22:00 on every national holiday requires a list to specify recurring time windows with different duration.

Reply:
I agree that the parameter name 'Filters' is kind of confusing. But the cron expression does serve as "filters" since we also have the parameter "Start" and "End" which should be considered as the timestamp when the filter starts to work("FilterStart") and when the filter ends working("FilterEnd").
The reason why we use "Filters" instead of "Recurrence" is that we will treat an empty list of "Filters" as not specified.
When "Filters" is not specifed, the TimeWindowFilter will revert to the original fixed time window filter.
The name "Filters" makes more sense.

Let's take a look at two choices:
Choice A:
"Start" : "2023-9-19-00:00",
"End": "2024-9-19-00:00",
"Filters": [ ]

Choice B:
"Start" : "2023-9-19-00:00",
"End": "2024-9-19-00:00",
"Recurrence": [ ]
If "Filters" or "Recurrence" is not specified, there is no confusion.
If "Filters" or "Recurrence" is not empty, the filter will only enable the feature during the recurring time window listed in the list "Filters" or "Recurrence". However, for the “Recurrence" choice, the parameter "Start" and "End" serve as the "RecurrenceStart" and "RecurrentEnd" instead of "FilterStart" and "FilterEnd". This will make the scenario that the "Recurrence" is specified as an empty list confusing. The function of "Start" and "End" will be inconsistent if we use name "Recurrence".
The user who uses FeatureManagement-Dotnet is expected to read the README before using it. The user who uses FeatureManagement through portal will never know the backend data schema for TimeWindowFilter unless they use "Advanced Edit"(in this case, the user should be considered as a pro who has read our doc). On the Portal, the scenarios of setting a fixed time window and setting recurring time window will have different UI and will not mention the "Filters" parameter.

For me, I will vote for the name "Filters" because it makes the role of "Start" and "End" parametes consistent. I also believe it will not cause confusion for the user who uses FeatureManagement-Dotnet lib directly.

Reply:
> The scenario when the user want to activate a feature flag during 18:00-20:00 on every weekend and during 18:00-22:00 on holidays requires a list to specify recurring time windows with different durations. 

I'm sure we can come up with even more complex scenarios. The question is how realistic they are. Outlook doesn't support infinite recurrence patterns, just one pattern at a time. Without the array, the parameters can be simplified to "Start", "End" and "RecurrencePattern". The UI design doesn't need an add button for more than one recurrence pattern anymore.

> Using multiple timewindow filters may not be a good choice, since the feature flag may also use other filters(e.g. targeting), in this case, the "RequirementType" is "All".

We know the "RequirementType" doesn't support all possible logical operations. It was a conscious decision. It's a balance between complexity and the most common scenarios. The choice of that design shouldn't be the reason for us to complicate the design of each feature filter. We have a reasonable workaround. We know there are scenarios that won't be supported out-of-box no matter how. That's where customers should consider custom filters.

Reply:
If we do not want to use list for "Recurrence", then we cannot use cron expression any more. The reason is that we have to use at least three cron expressions to describe the time span 18:30`20:30. e.g. "30-59 18 * * *", "* 19 * * *“ and "0-29 20 * * *".
I have proposed a lot of schema design to describe the recurrence. Some of them are deprecated early and are not included in the design doc.
The best way to describe recurrence is to specify the duration and recurrence pattern (i.e. when we want to make the recuurence happen).
At first, I want to use two parameters "Duration" and "RecurWhen". 
'''javascript
"TimeWindowDuration": { 
  "Type": "Hour", 
  "Length": 4 
},  


"RecurWhen": [
 "0 18 * * * 5"
]
```

Reply:
I'm not sure I follow what you said here. I did NOT suggest we abandon the usage of cron expression. I am only suggesting we support one cron expression per Time Window filter instead of an array of them. Here is an example.

``` JavaScript
"EnhancedPipeline": {
    "EnabledFor": [
        {
            "Name": "Microsoft.TimeWindow",
            "Parameters": {
                "Start": "2023-09-06T00:00:00+08:00",
                "RecurrencePattern": "* 18-19 * * Mon-Fri" // 18:00-20:00 on weekdays
            }
        }
    ]
}
```

Reply:
What if I want the recurrence happen during 18:30`20:30 every day?
You cannot represent 18:30-20:30 in a single cron.

Reply:
> > The scenario when the user want to activate a feature flag during 18:00-20:00 on every weekend and during 18:00-22:00 on holidays requires a list to specify recurring time windows with different durations.
> 
> I'm sure we can come up with even more complex scenarios. The question is how realistic they are. Outlook doesn't support infinite recurrence patterns, just one pattern at a time. Without the array, the parameters can be simplified to "Start", "End" and "RecurrencePattern". The UI design doesn't need an add button for more than one recurrence pattern anymore.
> 
> > Using multiple timewindow filters may not be a good choice, since the feature flag may also use other filters(e.g. targeting), in this case, the "RequirementType" is "All".
> 
> We know the "RequirementType" doesn't support all possible logical operations. It was a conscious decision. It's a balance between complexity and the most common scenarios. The choice of that design shouldn't be the reason for us to complicate the design of each feature filter. We have a reasonable workaround. We know there are scenarios that won't be supported out-of-box no matter how. That's where customers should consider custom filters.

Hi @zhenlan, I thinks scenario like `18:00-20:00 everyday` is a very common scenario, not something that rarely happens. And current design does not allow us to use one single cron to describe it. Since they are closed tied together, adding two cron expressions makes more sense to me than adding two TimeWindow filters.  If we don't support cron array, probablly customers will have to add more Feature filters, which means inceasing size/complexity in another level.

Reply:
@juniwang Yes, we chatted offline, and I understand this now. This limitation imposes challenges on our portal to properly break down `18:30-20:30` into 3 crontabs and reconstruct it back from crontabs to a human-readable time duration. We need to rethink our design choices.